### PR TITLE
[Merged by Bors] - feat : generalize PrimeSpectrum from Ring to Semiring

### PR DESCRIPTION
**Pull request by XavierXarles**
Some results of PrimeSpectrum generalized from CommRing to CommSemiring.

### CODE REVIEW

**Comment by XavierXarles** [comment]
> > And I think it would be nice if you can debug a bit further why some proofs become longer. Is there a mathematical reason, or is it because some API for semirings is under developed?
> 
> I am not sure to what proofs you refer as being longer. Do you mean the proof in AlgebraicGeometry.Properties of the instance?

Do I have to do something more?

File: Mathlib/AlgebraicGeometry/Properties.lean [hunk 1]
```diff
@@ -94,8 +94,9 @@ instance {R : CommRingCat} [H : _root_.IsReduced R] : IsReduced (Scheme.Spec.obj
   intro x; dsimp
   have : _root_.IsReduced (CommRingCat.of <| Localization.AtPrime (PrimeSpectrum.asIdeal x)) := by
     dsimp; infer_instance
-  exact isReduced_of_injective (StructureSheaf.stalkIso R x).hom
-    (StructureSheaf.stalkIso R x).commRingCatIsoToRingEquiv.injective
+  have :  ((Scheme.Spec.obj <| op R).toPresheafedSpace).presheaf.stalk x ≅
```
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/Properties.lean:97]: why is it now necessary to specify the type now?
```suggestion
  have : ((Scheme.Spec.obj <| op R).toPresheafedSpace).presheaf.stalk x ≅
```

**Review by alreadydone — COMMENTED** [review]
LGTM

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean [hunk 2]
```diff
@@ -617,21 +617,28 @@ theorem comap_injective_of_surjective (f : R →+* S) (hf : Function.Surjective
       (congr_arg PrimeSpectrum.asIdeal h : (comap f x).asIdeal = (comap f y).asIdeal))
 #align prime_spectrum.comap_injective_of_surjective PrimeSpectrum.comap_injective_of_surjective
 
-theorem comap_singleton_isClosed_of_surjective (f : R →+* S) (hf : Function.Surjective f)
-    (x : PrimeSpectrum S) (hx : IsClosed ({x} : Set (PrimeSpectrum S))) :
-    IsClosed ({comap f x} : Set (PrimeSpectrum R)) :=
+section CommRing
+
+variable (T : Type u) (T' : Type v) [CommRing T] [CommRing T']
```
Comment (eric-wieser) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:622]: Can you reuse the existing `R` and `S` here, by pushing the previous `CommSemiring` assumptions into `section`s?

Comment (XavierXarles) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:622]: > Can you reuse the existing `R` and `S` here, by pushing the previous `CommSemiring` assumptions into `section`s?

It is possible, but it involves more changes. Moreover, it is better then to reorganize some theorems, so the ones about CommRing are together. Do you like me to do it?

Comment (XavierXarles) [reply Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean]: I changed the file according to suggestions.

File: Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean [hunk 3]
```diff
@@ -718,6 +705,38 @@ theorem comap_inducing_of_surjective (hf : Surjective f) : Inducing (comap f) wh
...
+
+variable [CommRing R] [CommRing S]
+variable (f : R →+* S)
+variable {R}
+
+theorem comap_singleton_isClosed_of_surjective (f : R →+* S) (hf : Function.Surjective f)
+    (x : PrimeSpectrum S) (hx : IsClosed ({x} : Set (PrimeSpectrum S))) :
+    IsClosed ({comap f x} : Set (PrimeSpectrum R)) :=
+  haveI : x.asIdeal.IsMaximal := (isClosed_singleton_iff_isMaximal x).1 hx
+  (isClosed_singleton_iff_isMaximal _).2 (Ideal.comap_isMaximal_of_surjective f hf)
```
Comment (alreadydone) [inline Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean:724]: Once you generalize `Ideal.comap_isMaximal_of_surjective` you can probably generalize all results in this section that assumes surjectivity (not sure about IsIntegral). For surjectivity it should still be true for Semirings that there is an OrderIso between ideals in R containing the kernel and all ideals in S.

(Update: this is actually not true if R is only a semiring: consider R=Nat, S=Z/2Z and `f` the reduction mod 2 map. {0} is a maximal ideal whose preimage isn't (Nat\{1} is the unique maximal ideal).)

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**